File: api-chi/cmd/routes/movie_test.go
package routes

import (
	"api-chi/cmd/models"
	"api-chi/internal/message"

	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/go-chi/chi/v5"
	"github.com/stretchr/testify/assert"
)

func Test_MovieRoutes(t *testing.T) {
	r := chi.NewRouter()
	MovieRoutes(r)
	id := ""

	t.Run("Count success", func(t *testing.T) {
		req := httptest.NewRequest("GET", "/movie/count", nil)
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		response := message.Response{}
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.GET_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)
	})

	t.Run("GetAll success", func(t *testing.T) {
		search := ""
		limit := 10
		page := 1

		req := httptest.NewRequest("GET", fmt.Sprintf("/movie?search=%s&limit=%d&page=%d", search, limit, page), nil)
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		response := message.Response{}
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.GET_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)
	})

	t.Run("Create success", func(t *testing.T) {
		input := models.Movie{
			Name:            "test movie",
			InitialDate:     "2024-10-12",
			ThumbnailImage:  "test.png",
			TrailerVideoUrl: "test-trailer",
			Description:     "movie is test movie",
		}
		body, _ := json.Marshal(input)

		req := httptest.NewRequest("POST", "/movie", bytes.NewBuffer(body))
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		response := message.Response{}
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.CREATE_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)

		idValue, ok := response.Data.(string)
		if !ok {
			t.Fatalf("Expected response.Data to be of type string, got %T", response.Data)
		}
		id = idValue
	})

	t.Run("Update success", func(t *testing.T) {
		if id == "" {
			t.Fatal("ID must be set before running Update test")
		}

		input := models.Movie{
			Id:              id,
			Name:            "this is a only test movie",
			InitialDate:     "2024-12-30",
			ThumbnailImage:  "thumbnail-for-best-test-movie.png",
			TrailerVideoUrl: "this-best-trailer-video-url-for-test-movie",
			Description:     "this move is description for test only movie",
		}
		body, _ := json.Marshal(input)

		req := httptest.NewRequest("PATCH", "/movie", bytes.NewBuffer(body))
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		response := message.Response{}
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.UPDATE_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)
	})

	t.Run("Remove success", func(t *testing.T) {
		if id == "" {
			t.Fatal("ID must be set before running Remove test")
		}

		req := httptest.NewRequest("DELETE", "/movie/"+id, nil)
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		response := message.Response{}
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.REMOVE_DATA_SUCCESS, response.Message)
	})
}

File: api-fastify/src/tests/routes/movie.test.ts
import fastify from "fastify"
import MovieRoutes from "@/api/routes/movie"
import { Movie } from "@/types/movie"
import { HttpStatus } from "@/types/http_status"

describe("MovieRoutes", () => {
  let app: ReturnType<typeof fastify>
  let id: string = ""

  beforeAll(async () => {
    app = fastify()
    app.register(MovieRoutes)
    await app.ready()
  })

  afterAll(async () => {
    await app.close()
  })

  describe("POST /movies", () => {
    it("should create a movie director", async () => {
      const input: Movie = {
        id: "",
        name: "test movie is 1",
        initialDate: new Date(),
        thumbnailImage: "image.png",
        trailerVideoUrl: "trailer-url",
        description: "description about test movie is 1"
      }

      const response = await app.inject({
        method: "POST",
        url: "/movies",
        payload: input,
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      expect(typeof result.data.id).toBe("string")
      expect(typeof result.data.name).toBe("string")
      expect(result.data.name).toBe(input.name)
      id = result.data.id
    })
  })

  describe("PATCH /movies", () => {
    it("should update a movie director", async () => {
      const input: Movie = {
        id: id,
        name: "this is test movie 1",
        initialDate: new Date(),
        thumbnailImage: "test-movie-image.png",
        trailerVideoUrl: "test-movie-image-trailer",
        description: "test movie 1 description"
      }

      const response = await app.inject({
        method: "PATCH",
        url: "/movies",
        payload: input,
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      expect(typeof result.data.id).toBe("string")
      expect(typeof result.data.name).toBe("string")
      expect(result.data.id).toBe(input.id)
      expect(result.data.name).toBe(input.name)
    })
  })

  describe("DELETE /movies/:id", () => {
    it("should remove a movie director", async () => {
      const response = await app.inject({
        method: "DELETE",
        url: `/movies/${id}`,
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
    })
  })

  describe("GET /movies/count", () => {
    it("should count movie directors", async () => {
      const response = await app.inject({
        method: "GET",
        url: "/movies/count",
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      expect(typeof result.data).toBe("number")
    })
  })

  describe("GET /movies", () => {
    it("should get all movie directors", async () => {
      const response = await app.inject({
        method: "GET",
        url: "/movies?search=&limit=10&page=1",
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      let countData: number = 0
      result.data.forEach((movie: Movie) => {
        expect(typeof movie.id).toBe("string")
        expect(typeof movie.name).toBe("string")
        expect(movie.id).not.toBe("")
        expect(movie.name).not.toBe("")
        countData += 1
      })
      expect(countData).toBe(10)
    })
  })
})
